lib/gstdemuxeres.c

```diff
@@ -0,0 +1,614 @@
+/* DemuxerES
+ * Copyright (C) 2022 Igalia, S.L.
+ *     Author: Stephane Cerveau <scerveau@igalia.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You
+ * may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "gstdemuxeres.h"
+
+#include <gst/gst.h>
+
+static guint packet_counter = 0;
+GST_DEBUG_CATEGORY_STATIC (demuxer_es_debug);
+#define GST_CAT_DEMUXER_ES video_demuxer_es_debug
+
+
+typedef enum
+{
+  GST_AUTOPLUG_SELECT_TRY,
+  GST_AUTOPLUG_SELECT_EXPOSE,
+  GST_AUTOPLUG_SELECT_SKIP
+} GstAutoplugSelectResult;
+
+typedef enum _GstDemuxerESState
+{
+  /*< public > */
+  DEMUXER_ES_STATE_IDLE = 0,
+  DEMUXER_ES_STATE_READY,
+  DEMUXER_ES_STATE_ERROR,
+  DEMUXER_ES_STATE_EOS,
+} GstDemuxerESState;
+
+struct _GstDemuxerESPrivate
+{
+  GstElement *pipeline;
+
+  GAsyncQueue *packets;
+  GList *streams;
+
+  GstDemuxerESState state;
+  GCond ready_cond;
+  GMutex ready_mutex;
+  GMutex queue_mutex;
+};
+
+struct _GstDemuxerESPacketPrivate
+{
+  GstSample *sample;
+};
+
+static GstFlowReturn
+appsink_new_sample_cb (GstElement * sink, GstDemuxerEStream * stream)
+{
+  GstSample *sample;
+  GstBuffer *buffer;
+  GstDemuxerESPrivate *priv = stream->demuxer->priv;
+
+  g_signal_emit_by_name (sink, "pull-sample", &sample);
```

Comment:
why not to use the appsink callbacks rather than signals?

Reply:
bad habits I guess, better to use the callback you think ?

Reply:
it's less cpu intensive, afaik